src/openzeppelin/token/erc1155/library.cairo

```diff
@@ -0,0 +1,786 @@
+        }(account: felt, operator: felt) -> (approved: felt):
+        let (approved) = ERC1155_operator_approvals.read(account, operator)
+        return (approved)
+    end
+
+    #
+    # Externals
+    #
+
+    func set_approval_for_all{
+            syscall_ptr: felt*,
+            pedersen_ptr: HashBuiltin*,
+            range_check_ptr
+        }(operator: felt, approved: felt):
+        let (caller) = get_caller_address()
+        with_attr error_message("ERC1155: cannot approve from the zero address"):
+            assert_not_zero(caller)
+        end
+        _set_approval_for_all(caller, operator, approved)
+        return ()
+    end
+
+    func safe_transfer_from{
+            syscall_ptr: felt*,
+            pedersen_ptr: HashBuiltin*,
+            range_check_ptr
+        }(
+            from_: felt,
+            to: felt,
+            id: Uint256,
+            amount: Uint256,
+            data_len: felt,
+            data: felt*
+        ):
+        let (caller) = get_caller_address()
+        with_attr error_message("ERC1155: cannot call transfer from the zero address"):
+            assert_not_zero(caller)
+        end
+        with_attr error_message("ERC1155: caller is not owner nor approved"):
+            assert_owner_or_approved(from_)
+        end
+        _safe_transfer_from(from_, to, id, amount, data_len, data)
+        return ()
+    end
+
+    func safe_batch_transfer_from{
+            syscall_ptr: felt*,
+            pedersen_ptr: HashBuiltin*,
+            range_check_ptr
+        }(
+            from_: felt,
+            to: felt,
+            ids_len: felt,
+            ids: Uint256*,
+            amounts_len: felt,
+            amounts: Uint256*,
+            data_len: felt,
+            data: felt*
+        ):
+        let (caller) = get_caller_address()
+        with_attr error_message("ERC1155: cannot call transfer from the zero address"):
+            assert_not_zero(caller)
+        end
+        with_attr error_message("ERC1155: transfer caller is not owner nor approved"):
+            assert_owner_or_approved(from_)
+        end
+        _safe_batch_transfer_from(from_, to, ids_len, ids, amounts_len, amounts, data_len, data)
+        return ()
+    end
+
+    #
+    # Internals
+    #
+
+    func _safe_transfer_from{
+            syscall_ptr: felt*,
+            pedersen_ptr: HashBuiltin*,
+            range_check_ptr
+        }(
+            from_: felt,
+            to: felt,
+            id: Uint256,
+            amount: Uint256,
+            data_len: felt,
+            data: felt*
+        ):
+        alloc_locals
+        # Check args
+        with_attr error_message("ERC1155: transfer to the zero address"):
+            assert_not_zero(to)
+        end
+        with_attr error_message("ERC1155: id is not a valid Uint256"):
+            uint256_check(id)
+        end
+        with_attr error_message("ERC1155: amount is not a valid Uint256"):
+            uint256_check(amount)
+        end
```

Comment:
It is not computationally relevant but removing the amount check here also requires removing the `"ERC1155: insufficient balance for transfer"` message from the `sub_le` call underneath, since it can now fail for other reasons. This is ok, but it does mean that our errors are more generic. 

If performance is an issue, checking arithmetic manually (as in the Solidity implementation) would let us give specific errors, while also avoiding the redundant check of the initial balance (which cannot be an invalid uint256 supposing a correct implementation of the spec).